uptimerobot.py: fixed missing api actions

thieleon opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
Added all missing API actions that uptimerobot offers.

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

<!--- Please do not forget to include a changelog fragment:
      https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-changelog-fragments
      No need to include one for docs-only or test-only PR, and for new plugin/module PRs.
      Read about more details in CONTRIBUTING.md.
      -->
breaking_changes:
  - uptimerobot - requires api_key, action and optional params dictionary with additional parameters for the api call
##### ISSUE TYPE
<!--- Pick one or more below and delete the rest.
      'Test Pull Request' is for PRs that add/extend tests without code changes. -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the SHORT NAME of the module, plugin, task or feature below. -->
uptimerobot.py

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
The module was only able to `start` and `stop` monitors. With this pull request this was extended to all uptimerobot actions defined on [https://uptimerobot.com/api/](https://uptimerobot.com/api/).

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

thieleon commented:
The test fails as "requests" module is not installed on the sanity check. I didn't get how to do that though. If anyone can help me with that, I'll add this requirement to the file.

felixfontein reviewed (commented):
Thanks for your contribution! Please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment). Thanks.

Also please note that changes must be backwards compatibility. You cannot remove or rename parameters without making sure that the existing names still work.

thieleon commented:
@felixfontein I added the backwards compatibility using aliases and some code changes. Also added the changelog fragments. Let me know if anything is missing.

felixfontein commented:
> The test fails as "requests" module is not installed on the sanity check. I didn't get how to do that though. If anyone can help me with that, I'll add this requirement to the file.

Basically you have to adjust the PR to not use requests. Use ansible-core's `fetch_url` (as the old module does) instead of `requests`.

felixfontein commented on changelogs/fragments/7458-uptimerobot-add-missing-api-action.yml:
```diff
@@ -0,0 +1,2 @@
+major_changes:
+  - added all missing actions offered by the uptimerobot api (v2)
```
```suggestion
  - uptimerobot - added all missing actions offered by the uptimerobot v2 API (https://github.com/ansible-collections/community.general/pull/7458).
```

felixfontein commented on plugins/modules/uptimerobot.py:
```diff
@@ -7,16 +7,18 @@
 description:
-    - This module will let you start and pause Uptime Robot Monitoring
-author: "Nate Kingsley (@nate-kingsley)"
+    - This module will let you manage your Uptime Robot Monitoring
```
```suggestion
    - This module will let you manage your Uptime Robot Monitoring.
```

felixfontein commented on plugins/modules/uptimerobot.py:
```diff
@@ -25,133 +27,132 @@
         description:
-            - Define whether or not the monitor should be running or paused.
+            - Uptime Robot API key.
         required: true
```
```suggestion
        required: true
        aliases:
          - apikey
```

felixfontein commented on plugins/modules/uptimerobot.py:
```diff
@@ -25,133 +27,132 @@
         description:
-            - ID of the monitor to check.
+            - Uptime Robot API action.
         required: true
```
The `choices` list is missing here.

felixfontein commented on plugins/modules/uptimerobot.py:
```diff
@@ -25,133 +27,132 @@
-        required: true
+            - required data for api call. See https://uptimerobot.com/api/
+        required: false
 notes:
```
The `monitorid` option is not documented.

felixfontein commented on plugins/modules/uptimerobot.py:
```diff
@@ -25,133 +27,132 @@
+                ),
+                params=dict(type='dict', required=False),
+                # for backwards compatibility
+                state=dict(type='str', required=False),
```
`state` is already an alias of `action`. You cannot have `state` both as a separate action and as an alias.

felixfontein reviewed (commented):
I went once through the PR and marked a few things I spotted. But before I come to the details, here's some general thoughts:

Ansible modules should be state based and provide idempotency. (Look in particular at the last few points in https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#following-ansible-conventions.)

The old module already wasn't good at this (it always returns `changed=False`, and I'm not sure whether it actually passed if the current state was already there - that depends on how the old API worked).

The PR changes the module to a thin wrapper of a more complex API, so not providing much more than the [`uri` module](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/uri_module.html). This is usually not how Ansible modules should work.

thieleon commented:
@felixfontein I feel like using the API with the fetch_url seems to be a valid point here, as I didn't respect the idempotency. Once I have the time to properly implement it I will come back. I only pull-requested as the other module is  was not very useful for me.

felixfontein commented:
@thieleon ping!

needs_info
